internal/backend/grpc/resource.go

```diff
@@ -362,8 +363,13 @@ func CreateResource(resource *resources.Resource) (cosiresource.Resource, error)
 		return nil, err
 	}
 
+	unmarshalValue, err := initUnmarshalResourceSpec(obj)
+	if err != nil {
+		return nil, err
+	}
+
 	if resource.Spec != "" {
-		err = json.Unmarshal([]byte(resource.Spec), obj.Spec())
+		err = json.Unmarshal([]byte(resource.Spec), unmarshalValue)
```

Comment:
This approach worked for me in the infra provider:

```
		if r, ok := res.Spec().(interface {
			UnmarshalJSON(bytes []byte) error
		}); ok {
			if err = r.UnmarshalJSON([]byte(resource.Spec)); err != nil {
				return err
			}
		}
```

Reply:
hmm, this is cool, I'll try that.

Reply:
Couldn't make that work, as the type of `Spec()` here is:
`*protobuf.ResourceSpec[specs.ConfigPatchSpec,*specs.ConfigPatchSpec]`

This type implements `UnmarshalJSON`, but in the `protobuf` package, on `ResourceSpec`. And it simply ignores my overridden `UnmarshalJSON` function on the `ConfigPatchSpec`, as it uses `protojson`'s unmarshal function.

So what I need here is to call the `.Value` on the spec to get the actual `ConfigPatchSpec` object, and call its `UnmarshalJSON` function.

Am I missing something? Is there a neat way of doing it?

Reply:
Oh I see now. You have `UnmarshalJSON` defined on the wrapped spec.

I think the right way would be to customize COSI runtime code to call `UnmarshalJSON` on the wrapped spec if it implements it.
I'd really prefer to avoid the code with lots of reflection here.

Reply:
The change in the COSI runtime should be simple:
```go
// UnmarshalJSON implements json.Unmarshaler.
func (spec *ResourceSpec[T, S]) UnmarshalJSON(bytes []byte) error {
	spec.Value = new(T)

	if unmarshaler, ok := spec.(json.Unmarshaler); ok {
		return unmarshaler.UnmarshalJSON(bytes)
	}

	opts := protojson.UnmarshalOptions{}

	return opts.Unmarshal(bytes, spec.Value)
}
```